packages/core/src/domain/conversion/bundle-modifications/modifications.ts

```diff
@@ -0,0 +1,106 @@
+import { Bundle, Resource } from "@medplum/fhirtypes";
+import { cloneDeep } from "lodash";
+import * as uuid from "uuid";
+import { isPatient } from "../../../external/fhir/shared";
+import { DOC_ID_EXTENSION_URL } from "../../../external/fhir/shared/extensions/doc-id-extension";
+
+export type FhirExtension = {
+  url: string;
+  valueString: string;
+};
+
+export type FhirConverterParams = {
+  patientId: string;
+  fileName: string;
+  unusedSegments: string | undefined;
+  invalidAccess: string | undefined;
+};
+
+/**
+ * Replaces the IDs (and all references to them) for all resources that have an extension for the source document.
+ * This is done so that identical resources originated from different sources would have different IDs.
+ */
+export function replaceIdsForResourcesWithDocExtension(
+  fhirBundle: Bundle<Resource>,
+  patientId: string
+): Bundle<Resource> {
+  const updatedBundle = cloneDeep(fhirBundle);
+  const stringsToReplace: { old: string; new: string }[] = [];
+  if (!updatedBundle.entry || updatedBundle.entry.length === 0) {
+    throw new Error(`Missing bundle entries`);
+  }
+  for (const bundleEntry of updatedBundle.entry) {
+    if (!bundleEntry.resource) throw new Error(`Missing resource`);
+    if (!bundleEntry.resource.id) throw new Error(`Missing resource id`);
+    if (bundleEntry.resource.id === patientId) continue;
+
+    const resource = bundleEntry.resource;
+    if ("extension" in resource) {
+      const docIdExtension = resource.extension?.find(ext => ext.url === DOC_ID_EXTENSION_URL);
+      const idToUse = bundleEntry.resource.id;
+      const newId = uuid.v4();
```

Comment:
Any particular reason to use UUID v4 instead of v7, which is the one we chose because it's better for DB indexing?

Reply:
this function has been around for over a year. I tried to keep the changes minimal on this one. But ill update it, i guess

Reply:
I thought it was a new thing when I wrote that. I don't know if there are side effects of doing this. If you're not sure, it's fine to leave as is.

But if we know there's no side effects, then better change this sooner than later (we might want to store those on a FHIR server and the ID being a UUID v7 is helpful for performance reasons.